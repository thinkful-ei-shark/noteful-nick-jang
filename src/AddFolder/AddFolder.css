.add-folder {
  background-color: #030f1f;
  color: #b5b3b3;
  display: block;
  font-family: var(--custom-font);
  font-size: 14px;
  margin: 0 0 8px auto;
  max-width: 180px;
  padding: 8px 16px;
  text-decoration: none;
  line-height: 1em;
}

.add-folder p, .add-folder label {
  margin: 7px 0px;
}

.add-folder input {
  width: 100%;
  padding: 4px;
  margin: 7px;
  overflow: hidden;
}

.add-folder .form-title {
  font-size: 20px;
  font-style: normal;
}

.add-folder .hint {
  color: #FFB72F;
}

.add-folder .form-status {
  color: red;
}