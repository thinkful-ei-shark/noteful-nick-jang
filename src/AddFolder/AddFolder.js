import React, { Component } from 'react';
import config from '../config';
import ApiContext from '../ApiContext';
import ValidationError from '../ValidationError/ValidationError';
import PropTypes from 'prop-types';
import './AddFolder.css';

class AddFolder extends Component {
  static defaultProp = {
    id: '',
    deleteFolder: []
  }

  state = {
    folder: {
      value: '',
      touched: false
    },
    fetchError: ''
  }

  constructor() {
    super();
    this.folderRef = React.createRef();
  }

  updateFolder = (folder) => {
    this.setState({ folder: { value: folder, touched: true } });
  }

  validateFolder = () => {
    const folder = this.state.folder.value.trim();
    if (!folder) {
      if(this.state.folder.touched)
        this.folderRef.current.focus();
      return 'Name cannot be empty.';
    }
    return;
  }

  putFolder = (folderName) => {
    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ "name": folderName })
    }
    return fetch(`${config.API_ENDPOINT}/folders`, options)
      .then(res => {
        if (!res.ok) throw new Error(res.statusText);
        return res.json();
      })
  }

  handleSubmit = (event, callback) => {
    event.preventDefault();
    this.setState({ fetchError: '' })
    this.putFolder(this.state.folder.value)
      .then((resJson) => {
        this.props.deleteFolder(this.props.id);
        callback(resJson);
      })
      .catch(e => this.setState({ fetchError: e.message }));
  }

  componentDidMount() {
    if(this.state.folder.touched)
      this.folderRef.current.focus();
  }

  render() {
    return (
      <ApiContext.Consumer>
        {(context) => (
          <li className="add-folder">
            <form id='add-folder-form'>
              <p className='form-status'>{this.state.fetchError}</p>
              <p className='form-title'>Add Folder</p>
              <label htmlFor='folder'>Enter a folder name:</label>
              <p className='hint'>* required</p>
              <input
                ref={this.folderRef}
                type='text'
                id='folder'
                name='folder'
                aria-required="true"
                aria-describedby='folder-description'
                aria-label="Enter a folder name"
                aria-invalid={!!this.validateFolder()}
                onChange={(e) => this.updateFolder(e.target.value)}
              />
              {this.state.folder.touched && <ValidationError id='folder-description' message={this.validateFolder()} />}
              <button
                type='submit'
                form='add-folder-form'
                onClick={(e) => {
                  this.handleSubmit(e, context.addFolder)
                }}
                disabled={this.validateFolder()}
              > Submit
            </button>
              <button
                type='reset'
                form='add-folder-form'
                onClick={(e) => this.props.deleteFolder(this.props.id, e)}
              > Cancel
              </button>
            </form>
          </li>
        )}
      </ApiContext.Consumer>
    );
  }
}

AddFolder.propTypes = {
  deleteFolder: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired
}

export default AddFolder;