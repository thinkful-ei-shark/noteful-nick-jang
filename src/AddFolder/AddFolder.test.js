import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import AddFolder from './AddFolder'

describe(`AddFolder component`, () => {
  it('renders AddFolder by default', () => {
    const wrapper = shallow(<AddFolder />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
