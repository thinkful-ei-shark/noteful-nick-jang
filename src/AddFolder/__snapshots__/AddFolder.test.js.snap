// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddFolder component renders AddFolder by default 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
